src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,302 @@
+            _projectFileWatcherFactory = projectFileModelWatcherFactory;
+            _textBufferListenerFactory = textBufferListenerFactory;
+            _textBufferManager = textBufferManager;
+        }
+
+        #region Initialization/Destruction
+
+        /// <summary>
+        /// Called by anything attempting to open a project file editor window, usually by a command. This will show the window frame, creating it if
+        /// has not already been created.
+        /// </summary>
+        public async Task OpenEditorAsync()
+        {
+            await _threadingService.SwitchToUIThread();
+            lock (_lock)
+            {
+                // If the editor is already open, just show it and return
+                if (_currentState != EditorState.NoEditor)
+                {
+                    // If we're initializing, _windowFrame might be null. In that case, when the initialization code
+                    // is done, it'll take care of showing the frame.
+                    _windowFrame?.Show();
+                    return;
+                }
+                _currentState = EditorState.Initializing;
+            }
+
+            // First, open the frame
+            var loadedProjectEditorGuid = Guid.Parse(LoadedProjectFileEditorFactory.EditorFactoryGuid);
+            _windowFrame = _shellHelper.OpenDocumentWithSpecificEditor(_serviceProvider, _unconfiguredProject.FullPath, loadedProjectEditorGuid, Guid.Empty);
+
+            // Set up the project file watcher, so changes to the project file are detected and the buffer is updated.
+            _projectFileModelWatcher = _projectFileWatcherFactory.CreateExport();
+            _projectFileModelWatcher.Initialize();
+
+            // Finally, show the frame and move to EditorClean in lockstep
+            lock (_lock)
+            {
+                _windowFrame.Show();
+                _currentState = EditorState.EditorClean;
+                _projectFileModelWatcher.Initialize();
+            }
+        }
+
+        /// <summary>
+        /// Called from the editor wrapper. This sets up the listener that controls the dirty state and saving commands.
+        /// </summary>
+        public async Task InitializeTextBufferStateListenerAsync(WindowPane hostPane)
+        {
+            _textBufferStateListener = _textBufferListenerFactory.CreateExport();
+            await _textBufferStateListener.InitializeAsync(hostPane).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Starts the process of closing the project file editor, if one is open currently.
+        /// </summary>
+        public async Task CloseWindowAsync()
+        {
+            lock (_lock)
+            {
+                if (_currentState == EditorState.NoEditor) return;
+
+                // If there was no changes, then we just close the editor
+                if (_currentState == EditorState.EditorClean || _currentState == EditorState.ProjectFileChangingFromClean)
```

Comment:
What's the relevance of the ProjectFileChangingFromClean state when compared to EditorClean? You don't seem to do anything different with this state.